conduit/pipeline.go

```diff
@@ -0,0 +1,286 @@
+package conduit
+
+import (
+	"context"
+	"fmt"
+	"github.com/algorand/go-algorand-sdk/encoding/json"
+	"github.com/algorand/indexer/plugins"
+
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v3"
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/algorand/go-algorand/util"
+	"github.com/algorand/indexer/data"
+	"github.com/algorand/indexer/exporters"
+	"github.com/algorand/indexer/importers"
+	"github.com/algorand/indexer/processors"
+)
+
+const autoLoadParameterConfigName = "conduit.yml"
+
+// PipelineConfig stores configuration specific to the conduit pipeline
+type PipelineConfig struct {
+	ConduitConfig *Config
+
+	PipelineLogLevel string `yaml:"LogLevel"`
+	// Store a local copy to access parent variables
+	Importer   map[string]interface{}   `yaml:"Importer"`
+	Processors []map[string]interface{} `yaml:"Processors"`
+	Exporter   map[string]interface{}   `yaml:"Exporter"`
+}
+
+// Valid validates pipeline config
+func (cfg *PipelineConfig) Valid() error {
+	if cfg.ConduitConfig == nil {
+		return fmt.Errorf("conduit configuration was nil")
+	}
+
+	if _, err := log.ParseLevel(cfg.PipelineLogLevel); err != nil {
+		return fmt.Errorf("pipeline log level (%s) was invalid: %w", cfg.PipelineLogLevel, err)
+	}
+
+	if len(cfg.Importer) == 0 {
+		return fmt.Errorf("importer configuration was empty")
+	}
+
+	if len(cfg.Exporter) == 0 {
+		return fmt.Errorf("exporter configuration was empty")
+	}
+
+	return nil
+}
+
+// MakePipelineConfig creates a pipeline configuration
+func MakePipelineConfig(logger *log.Logger, cfg *Config) (*PipelineConfig, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("empty conduit config")
+	}
+
+	// double check that it is valid
+	if err := cfg.Valid(); err != nil {
+		return nil, err
+	}
+
+	pCfg := PipelineConfig{PipelineLogLevel: logger.Level.String(), ConduitConfig: cfg}
+
+	// Search for pipeline configuration in data directory
+	autoloadParamConfigPath := filepath.Join(cfg.ConduitDataDir, autoLoadParameterConfigName)
+	paramConfigFound := util.FileExists(autoloadParamConfigPath)
+
+	if !paramConfigFound {
+		return nil, fmt.Errorf("could not find %s in data directory (%s)", autoLoadParameterConfigName, cfg.ConduitDataDir)
+	}
+
+	logger.Infof("Auto-loading Conduit Configuration: %s", autoloadParamConfigPath)
+
+	configStr, err := ioutil.ReadFile(autoloadParamConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	err = yaml.Unmarshal(configStr, &pCfg)
```

Comment:
If we are using viper, we should use it here also.

Reply:
Would viper read the file in?

Reply:
Yes, you give it a path and it loads in the config.

Looks like it has utilities for extracting subtrees too:
https://github.com/spf13/viper#extracting-a-sub-tree